File: src/main/java/me/zeroeightsix/kami/module/modules/misc/AutoReconnect.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.ScreenEvent;
import me.zeroeightsix.kami.mixin.client.IDisconnectedScreen;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.client.gui.screen.ConnectScreen;
import net.minecraft.client.gui.screen.DisconnectedScreen;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.options.ServerEntry;

/**
 * Created by 086 on 9/04/2018.
 */
@Module.Info(name = "AutoReconnect", description = "Automatically reconnects after being disconnected", category = Module.Category.MISC, alwaysListening = true)
public class AutoReconnect extends Module {

    private Setting<Integer> seconds = register(Settings.integerBuilder("Seconds").withValue(5).withMinimum(0).build());
    private static ServerEntry cServer;

    @EventHandler
    public Listener<ScreenEvent.Closed> closedListener = new Listener<>(event -> {
        if (event.getScreen() instanceof ConnectScreen)
            cServer = mc.getCurrentServerEntry();
    });

    @EventHandler
    public Listener<ScreenEvent.Displayed> displayedListener = new Listener<>(event -> {
        if (isEnabled() && event.getScreen() instanceof DisconnectedScreen && (cServer != null || mc.getCurrentServerEntry() != null))
            event.setScreen(new KamiDisconnectedScreen((DisconnectedScreen) event.getScreen()));
    });

    private class KamiDisconnectedScreen extends DisconnectedScreen {
        private Screen parent;
        int millis = seconds.getValue() * 1000;
        long cTime;

        public KamiDisconnectedScreen(DisconnectedScreen disconnected) {
            super(((IDisconnectedScreen) disconnected).getParent(), disconnected.getTitle().asString(), ((IDisconnectedScreen) disconnected).getReason());
            cTime = System.currentTimeMillis();
            parent = ((IDisconnectedScreen) disconnected).getParent();
        }

        @Override
        public void tick() {
            if (millis <= 0)
                mc.openScreen(new ConnectScreen(parent, mc, cServer == null ? mc.getCurrentServerEntry() : cServer));
        }

        @Override
        public void render(int mouseX, int mouseY, float partialTicks) {
            super.render(mouseX, mouseY, partialTicks);

            long a = System.currentTimeMillis();
            millis -= a - cTime;
            cTime = a;

            String s = "Reconnecting in " + Math.max(0, Math.floor((double) millis / 100) / 10) + "s";
            font.drawWithShadow(s, width / 2 - font.getStringWidth(s) / 2, height - 16, 0xffffff);
        }

    }

}

File: src/main/kotlin/me/zeroeightsix/kami/module/modules/AutoReconnect.kt
package me.zeroeightsix.kami.module.modules

import me.zero.alpine.listener.EventHandler
import me.zero.alpine.listener.EventHook
import me.zero.alpine.listener.Listener
import me.zeroeightsix.kami.event.events.ScreenEvent
import me.zeroeightsix.kami.event.events.ScreenEvent.Displayed
import me.zeroeightsix.kami.mixin.client.IDisconnectedScreen
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import net.minecraft.client.gui.screen.ConnectScreen
import net.minecraft.client.gui.screen.DisconnectedScreen
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.options.ServerEntry
import kotlin.math.floor

/**
 * Created by 086 on 9/04/2018.
 */
@Module.Info(
    name = "AutoReconnect",
    description = "Automatically reconnects after being disconnected",
    category = Module.Category.MISC,
    alwaysListening = true
)
object AutoReconnect : Module() {

    private var cServer: ServerEntry? = null
    private val seconds = register(Settings.integerBuilder("Seconds").withValue(5).withMinimum(0).build())
    @EventHandler
    var closedListener =
        Listener(
            EventHook { event: ScreenEvent.Closed ->
                if (event.screen is ConnectScreen) cServer =
                    mc.currentServerEntry
            }
        )

    @EventHandler
    var displayedListener = Listener(
        EventHook { event: Displayed ->
            if (isEnabled() && event.screen is DisconnectedScreen && (cServer != null || mc.currentServerEntry != null)) event.screen =
                KamiDisconnectedScreen(event.screen as DisconnectedScreen)
        }
    )

    private class KamiDisconnectedScreen(disconnected: DisconnectedScreen) : DisconnectedScreen(
        (disconnected as IDisconnectedScreen).parent,
        disconnected.title.asString(),
        (disconnected as IDisconnectedScreen).reason
    ) {
        private val parent: Screen
        var millis = seconds.value.toLong() * 1000
        var cTime: Long

        override fun tick() {
            if (millis <= 0) mc.openScreen(
                ConnectScreen(
                    parent,
                    mc,
                    if (cServer == null) mc.currentServerEntry else cServer
                )
            )
        }

        override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
            super.render(mouseX, mouseY, partialTicks)
            val a = System.currentTimeMillis()
            millis -= (a - cTime.toInt())
            cTime = a
            val s = "Reconnecting in " + 0.0.coerceAtLeast(floor(millis.toDouble() / 100) / 10) + "s"
            font.drawWithShadow(s, width / 2 - font.getStringWidth(s) / 2.toFloat(), height - 16.toFloat(), 0xffffff)
        }

        init {
            cTime = System.currentTimeMillis()
            parent = (disconnected as IDisconnectedScreen).parent
        }

    }
}